README.md
REST API backend for vue-storefront
===================================

This is a backend service for [vue-storefront](https://github.com/DivanteLtd/vue-storefront). Provides data access to product catalog (via ElasticSearch) and allows users to place orders into order queue (by default it's Redis queue supported via kqueue library).

## Vue Storefront

Vue Storefront is a standalone [PWA](https://developers.google.com/web/progressive-web-apps/) (Progressive Web Application ) storefront for your eCommerce, possible to connect with any eCommerce backend (eg. Magento, Prestashop or Shopware) through the API.

Vue Storefront is and always will be in the open source. Anyone can use and support the project, we want it to be a tool for the improvement of the shopping experience. The project is still in the prove of concept phase. We are looking for Contributors and Designer willing to help us the the solution development.

Vue Storefront was build as a all-in-one front-end for eCommerce. For providing the best performance we decided to use Vue.js as a front-end library, Node.js + Express (and maybe GraphQL support) as a server-API, Elastic Search as a database of products and full PWA/off-line support.
Here you can read more about the proof of concept for [Vue Storefront connected with Magento2](https://www.linkedin.com/pulse/magento2-nosql-database-pwa-support-piotr-karwatka).

Besides a big improvement for the shopping experience, we also want to create a great code base for every developer who needs to work on a front-end application for the eCommerce.

## Requirements

- Docker and Docker Compose

Already included in `vue-storefront-api` Docker image (required locally, if you do not use containerization):
- Node.js 10.x or higher
- Yarn

## Installation

**Start a containerized environment**

The **legacy** (A) mode - starting just the Elastic and Redis containers:
`docker-compose up -d`

The **standard** (B) mode - starting Elastic, Redis + Vue Storefront API containers:
`docker-compose -f docker-compose.yml -f docker-compose.nodejs.yml up -d`

As a result, all necessary services will be launched:
- Vue Storefront API runtime environment (Node.js with dependencies from `package.json`)
- [ElasticSearch](https://www.elastic.co/products/elasticsearch)
- [Redis](https://redis.io/)
- Kibana (optional)

**Import product catalog**

Product catalog is imported using [elasticdump](https://www.npmjs.com/package/elasticdump), which is installed automatically via project dependency. The default ElasticSearch index name is: `vue_storefront_catalog`

- (A) `yarn restore`
- (B) `docker exec -it vue-storefront-api_app_1 yarn restore`

Then, to update the structures in the database to the latest version (data migrations), do the following:

- (A) `yarn migrate`
- (B) `docker exec -it vue-storefront-api_app_1 yarn migrate`

By default, the application server is started in development mode. It means that code auto reload is enabled along with ESLint, babel support.

It restores JSON documents stored in `./var/catalog.json`. The opposite command - used to generate `catalog.json` file from running ElasticSearch cluster:

- (A) `yarn dump`
- (B) `docker exec -it vue-storefront-api_app_1 yarn dump`

**Access ElasticSearch data with Kibana**

A [Kibana](https://www.elastic.co/products/kibana) service is available to explore, search and visualize indexed data at the following url:

`http://localhost:5601/`

At first access it will ask to specify an index pattern, insert `vue_storefront*`

**Note:** Kibana is not provided with the Elastic 7 docker file. Please install it on your own when needed or check the [`es-head`](https://chrome.google.com/webstore/detail/elasticsearch-head/ffmkiejjmecolpfloofpjologoblkegm) Chrome plugin.

## Elastic 7 support

By default, Vue Storefront API docker files and config are based on Elastic 5.6. We plan to change the default Elastic version to 7 with the 1.11 stable release. As for now, the [Elastic 7 support](https://github.com/DivanteLtd/vue-storefront-api/pull/342) is marked as **experimental**. 

**How to use it?**

In order to use the new Elastic 7 please make sure your `vue-storefront-api` has been checked out from the [PR#342](https://github.com/DivanteLtd/vue-storefront-api/pull/342) branch.
To start the Elastic 7 docker service please use the `docker-compose.elastic7.yml` file provided:

```bash
docker-compose -f docker-compose.elastic7.yml up
```

Then, please do change the `config/local.json` to start using the new Elastic API. Key properties in the `elasticsearch` section are: `indexTypes` and `apiVersion` (to be set to 7.1). If you're using the multistore configuration please make sure you adjusted the `storeViews.*.elasticsearch` section as well - per each separate store.

```json
  "elasticsearch": {
    "host": "localhost",
    "index": "vue_storefront_catalog",
    "port": 9200,
    "protocol": "http",
    "min_score": 0.01,
    "indices": [
      "vue_storefront_catalog",
      "vue_storefront_catalog_de",
      "vue_storefront_catalog_it"
    ],
    "indexTypes": [
      "product",
      "category",
      "cms_block",
      "cms_page",
      "attribute",
      "taxrule",
      "review"
    ],
    "apiVersion": "7.1"
  }
```

Starting from [Elasitc 6 and 7](https://www.elastic.co/guide/en/elasticsearch/reference/current/breaking-changes-7.0.html) we can have **just single** document type per single index. Vue Storefront used to have `product`, `category` ... types defined in the `vue_storefront_catalog`.

From now on, we're using the separate indexes per each entity type. The convention is: `${indexName}_${entityType}`. If your' **logical index name** is `vue_storefront_catalog` then it will be mapped to the **physical indexes** of: `vue_storefront_catalog_product`, `vue_storefront_catalog_category` ...

[Tools](https://docs.vuestorefront.io/guide/data/database-tool.html) are adjusted to ES7. You can use `yarn db new`, `yarn restore`, `yarn mage2vs import`. Just make sure that you have set up `config.elasticsearch.apiVersion` to `7.1`.

## API access
Catalog API calls are compliant with ElasticSearch (it works like a filtering proxy to ES). More on ES queries: [ElasticSearch queries tutorial](http://okfnlabs.org/blog/2013/07/01/elasticsearch-query-tutorial.html)

Elastic search endpoint: `http://localhost:8080/api/catalog/search/<INDEX_NAME>/`. You can run the following command to check if everything is up and runing (it assumes `vue_storefront_catalog` as default index name):

`curl -i http://elastic:changeme@localhost:8080/api/catalog/vue_storefront_catalog/_search`

## Data formats
This backend is using ElasticSearch data formats popularized by [ElasticSuite for Magento2 by Smile.fr](https://github.com/Smile-SA/elasticsuite).

* [Product model](./src/models/catalog-product.md)
* [Category model](./src/models/catalog-category.md)

## Data migrations
Please use data migration mechanism provided to manipulate Redis, ElasticSearch or kue. Details: https://github.com/DivanteLtd/vue-storefront-api/tree/master/doc 

## Adding custom modules with own dependencies (Yarn only)
When adding custom [Extensions to the API](https://github.com/DivanteLtd/vue-storefront/blob/master/doc/Extending%20vue-storefront-api.md) you might want to define some dependencies inside them. Thanks to [Yarn workspaces](https://yarnpkg.com/lang/en/docs/workspaces/) dependencies defined inside your custom module will be installed when you execute `yarn` at project root level, so it's way easier and faster than installing all modules dependencies separately.

To do this, define the `package.json` with your dependencies in your custom module:
- `src/api/extensions/{your-custom-extension}/package.json` 
- `src/platforms/{your-custom-platform}/package.json`

Executing `docker exec -it vue-storefrontapiapp_1 yarn install` will also download your custom modules dependencies.

NOTE: `npm` users will still have to install the dependencies individually in their modules.

## Reviews
To use review feature you need to install custom module for Magento 2: [Divante ReviewApi](https://github.com/DivanteLtd/magento2-review-api)
By default new reviews will be added with status "Pending". 
```json
  "review": {
    "defaultReviewStatus": 2 
  },
```

## Output Cache
Vue Storefront API supports output cache for catalog operations. Cache is tagged and can by dynamically invalidated. Please find the details how to configure it [in our docs](https://docs.vuestorefront.io/guide/basics/ssr-cache.html).

You can manually clear the Redis cache for specific tags by running the following command:

```bash
npm run cache clear
npm run cache clear -- --tag=product,category
npm run cache clear -- --tag=P198
npm run cache clear -- --tag=*
```

## Running initial Magento2 import

Magento2 data import is now integrated into `vue-storefront-api` for simplicity. It's still managed by the [mage2vuestorefront](https://github.com/DivanteLtd/mage2vuestorefront) - added as a dependency to `vue-storefront-api`.

After setting the `config.magento2.api` section using Yours Magento2 oauth credentials:

```json
  "magento2": {
    "imgUrl": "http://localhost:8080/media/catalog/product",
    "assetPath": "/../var/magento2-sample-data/pub/media",
    "api": {
      "url": "http://demo-magento2.vuestorefront.io/rest",
      "consumerKey": "byv3730rhoulpopcq64don8ukb8lf2gq",
      "consumerSecret": "u9q4fcobv7vfx9td80oupa6uhexc27rb",
      "accessToken": "040xx3qy7s0j28o3q0exrfop579cy20m",
      "accessTokenSecret": "7qunl3p505rubmr7u1ijt7odyialnih9"
    }
  },
```

You can run the following command to execute the full import:

```bash
 yarn mage2vs import
 ```

 ... or in multistore setup You can run the same command with specified `store-code` parameter

```bash
 yarn mage2vs import --store-code=de
 ```

 Import of CMS blocks and pages is also performed by the full import. If the CMS API extension ((SnowdogApps/magento2-cms-api)[https://github.com/SnowdogApps/magento2-cms-api]) was not installed in magento 2, it's recommended to skip the CMS import commands.

```bash
 yarn mage2vs import --skip-blocks --skip-pages
 ```
 
## Executing delta indexer

You can use the following command to run a delta indexer for a specific storeview:

```
yarn mage2vs productsdelta --store-code=de
```

## Self signed certificates

Often in non production environment other services are using self signed certificates for secure connection.
You can easily setup the application to trust them by putting them in config/certs directory.  

License
-------

[MIT](./LICENSE)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.12.4] - Unreleased

### Added
- Added webp support for local image action - @gibkigonzo (#409)

### Fixed 

 * add try catch to jwt.decode for groupId which is needed when you change the key for the JWT - @resubaka (#496)
 * Explicit check if the msi stock id is set for the given storeview  - [@philippsander](https://github.com/philippsander) (#483)

### Added

 - Add extended ElasticSearch 7+ support and more flexibility to `elastic` lib - @cewald (#512)
 
## [1.12.3] - 2020.07.23

### Fixed

- Bump version for `vsf-utilities` - @gibkigonzo (#495)
- Responsing with error instead of throwing for broken /api/catalog paths - @Fifciu

## [1.12.2] - 2020.07.20

### Added

- Added validation for user profile update. - @gibkigonzo (#488) - **Breaking change!** - from this change you can't send anything when you update user data. You can only send what is specified in `src/models/userProfileUpdate.schema.json`. If you still want to send anything then change `"additionalProperties": false,` to `true`.
- add `getToken` to handle getting token from header - @gibkigonzo (#488)
- Allow filter elasticsearch request for security reasons - @seSze (#476)

### Fixed

- Fix default value for `maxAgeForResponse` - @lauraseidler (#485)
- Adds `vsf-utlilities`. Use productEquals to compare products in o2m - @gibkigonzo (#477)

## [1.12.1] - 2020.06.22

### Added

- Add request_cache to ES url - @gibkigonzo (#387)

### Fixed

- Add error status code as number in `apiError` - @gibkigonzo (#442)
- Get proper tax calculation for multistore - @didkan (#464)
- Create only one ES client instance per app - @gibkigonzo (#393)

## [1.12.0] - 2020.06.01

### Added
- Add url module - @gibkigonzo (#3942)
- The `response_format` query parameter to the `/api/catalog` endpoint. Currently there is just one additional format supported: `response_format=compact`. When used, the response format got optimized by: a) remapping the results, removing the `_source` from the `hits.hits`; b) compressing the JSON fields names according to the `config.products.fieldsToCompact`; c) removing the JSON fields from the `product.configurable_children` when their values === parent product values; overall response size reduced over -70% - @pkarw
- The support for `SearchQuery` instead of the ElasticSearch DSL as for the input to `/api/catalog` - using `storefront-query-builder` package - @pkarw - https://github.com/DivanteLtd/vue-storefront/issues/2167
- Create attribute service that allows to fetch attributes with specific options - used for products aggregates - @gibkigonzo (https://github.com/DivanteLtd/vue-storefront/pull/4001, https://github.com/DivanteLtd/mage2vuestorefront/pull/99)
- Add ElasticSearch client support for HTTP authentication - @cewald (#397)
- Endpoint for reset password with reset token. Only for Magento 2 - @Fifciu
- Varnish Cache with autoinvalidation by Cache tags as addon - @Fifciu
- Add `resetPasswordUsingResetToken` to `magento1` platform - @cewald (#415)
- Update to `storefront-query-builder` version `1.0.0` - @cewald (#429)
- Add `composeError` in './magento2/o2m.js' - @flancer64 (#422)
- Explicit data extraction from 'Error' objects - @flancer64 (#424)
- Merge scripts: db7 => db, elastic7=>elastic and add support for es7 in yarn mage2vs import - @gibkigonzo (#436)

### Fixed
- add es7 support for map url module and fixed default index for es config - @gibkigonzo
- Add correct paths for production build - @cewald (#407)
- Fix MSI default stock id value
- Add outputFormatter to response from cache - @gibkigonzo (#428)
- disable showing stack for invalid requests - @gibkigonzo (#431)
- Improve `_outputFormatter` on cache catalog-response to prevent exception - @cewald (#432)
- use ts for compiling additional scripts - @gibkigonzo (#437)
- Bugfix for wrong JSON scheme url's and hostname resolution for `magento1` platform - @cewald (#443)
- check if headers has been already sent in image middleware - @gibkigonzo (#434)
- Missing packages in Docker Makefile (make, python and g++) - @proxiblue (#448)
- change path in ecosystem for o2m - @gibkigonzo (#434)
- udpate typings for query params - @gibkigonzo (#446)

## [1.11.1] - 2020.03.17

### Added
- Add save address on place order - @lucasqm (#394)
- Add ElasticSearch client support for HTTP authentication - @cewald (#397)
- Add error handling for catalog and add header 'X-VS-Cache-Tags' to response - @gibkigonzo

### Fixed
- Add fallback for `sourcePriceInclTax` and `finalPriceInclTax` in `magento1` platform - @cewald (#398)
- Add default ES index to config and update `getStockList` - @gibkigonzo (#405)
- Makes elastic-stock extension compatible with both ES5 and ES7. Allows for stock fetch of configurable children that is set as "Not Visible Individually" - @didkan (#410)


## [1.11.0] - 2019.12.20

### Fixed

- Fixed some smaller issues with graphql so that it is now working again with the fronted - #350
- Replaced the old `crop` function call which has been removed from Sharp image processor - @grimasod (#381)
- Add product processor to new URL mapper endpoint #401 - @cewald (#401, #403)
- Add fallback for `sourcePriceInclTax` and `finalPriceInclTax` in `magento1` platform - @cewald (#398)

## [1.11.0-rc.1] - 2019.10.03

### Added
- Experimental Elastic 7 support - @pkarw (#342)
- Output cache support with tagging and cache invalidate requests forwarding - @pkarw @resubaka (https://github.com/DivanteLtd/vue-storefront/issues/3367, #333)
- Constant for Mailchimp subscription status - @KonstantinSoelch (#294)
- mage2vs import now has optional `--generate-unique-url-keys` parameter which defaults to `false` to enable/disable the url key generation with name and id for categories - @rain2o (#232)

src/lib/elastic.js
const path = require('path')
const _ = require('lodash')
const fs = require('fs');
const jsonFile = require('jsonfile')
const es = require('@elastic/elasticsearch')
const querystring = require('querystring')

function adjustIndexName (indexName, entityType, config) {
  if (parseInt(config.elasticsearch.apiVersion) < 6) {
    return indexName
  } else {
    return `${indexName}_${entityType}`
  }
}

function decorateBackendUrl (entityType, url, req, config) {
  const {
    useRequestFilter,
    requestParamsBlacklist,
    overwriteRequestSourceParams,
    apiVersion
  } = config.elasticsearch

  if (useRequestFilter && typeof config.entities[entityType] === 'object') {
    const urlParts = url.split('?')
    const { includeFields, excludeFields } = config.entities[entityType]

    const filteredParams = Object.keys(req.query)
      .filter(key => !requestParamsBlacklist.includes(key))
      .reduce((object, key) => {
        object[key] = req.query[key]
        return object
      }, {})

    let _source_include = includeFields || []
    let _source_exclude = excludeFields || []

    if (!overwriteRequestSourceParams) {
      const requestSourceInclude = req.query._source_include || []
      const requestSourceExclude = req.query._source_exclude || []
      _source_include = [...includeFields, ...requestSourceInclude]
      _source_exclude = [...excludeFields, ...requestSourceExclude]
    }

    const isEs6AndUp = (parseInt(apiVersion) >= 6)
    let _sourceIncludeKey = isEs6AndUp ? '_source_includes' : '_source_include'
    let _sourceExcludeKey = isEs6AndUp ? '_source_excludes' : '_source_exclude'

    const urlParams = {
      ...filteredParams,
      [_sourceIncludeKey]: _source_include,
      [_sourceExcludeKey]: _source_exclude
    }
    url = `${urlParts[0]}?${querystring.stringify(urlParams)}`
  }

  return url
}

function adjustQueryParams (query, entityType, config) {
  delete query.request
  delete query.request_format
  delete query.response_format

  const {
    apiVersion,
    useRequestFilter,
    overwriteRequestSourceParams,
    requestParamsBlacklist,
    cacheRequest
  } = config.elasticsearch

  if (useRequestFilter && !overwriteRequestSourceParams && typeof config.entities[entityType] === 'object') {
    let { includeFields, excludeFields } = config.entities[entityType]
    const requestSourceInclude = query._source_include ? query._source_include.split(',') : []
    const requestSourceExclude = query._source_exclude ? query._source_exclude.split(',') : []
    query._source_include = [...includeFields, ...requestSourceInclude]
    query._source_exclude = [...excludeFields, ...requestSourceExclude]
  }

  if (parseInt(apiVersion) >= 6) { // legacy for ES 5
    query._source_includes = query._source_include
    query._source_excludes = query._source_exclude
    delete query._source_exclude
    delete query._source_include
    if (cacheRequest) {
      query.request_cache = !!cacheRequest
    }
  }

  if (useRequestFilter && typeof config.entities[entityType] === 'object') {
    query = Object.keys(query)
      .filter(key => !requestParamsBlacklist.includes(key))
      .reduce((object, key) => {
        object[key] = query[key]
        return object
      }, {})
  }

  return query
}

function adjustBackendProxyUrl (req, indexName, entityType, config) {
  let url
  const queryString = require('query-string');
  const parsedQuery = adjustQueryParams(queryString.parseUrl(req.url).query, config)

  if (parseInt(config.elasticsearch.apiVersion) < 6) { // legacy for ES 5
    url = config.elasticsearch.host + ':' + config.elasticsearch.port + '/' + indexName + '/' + entityType + '/_search?' + queryString.stringify(parsedQuery)
  } else {
    url = config.elasticsearch.host + ':' + config.elasticsearch.port + '/' + adjustIndexName(indexName, entityType, config) + '/_search?' + queryString.stringify(parsedQuery)
  }

  if (!url.startsWith('http')) {
    url = config.elasticsearch.protocol + '://' + url
  }

  return decorateBackendUrl(entityType, url, req, config)
}

function adjustQuery (esQuery, entityType, config) {
  if (parseInt(config.elasticsearch.apiVersion) < 6) {
    esQuery.type = entityType
  } else {
    delete esQuery.type
  }

  esQuery.index = adjustIndexName(esQuery.index, entityType, config)
  return esQuery
}

function getHits (result) {
  if (result.body) { // differences between ES5 and ES7
    return result.body.hits.hits
  } else {
    return result.hits.hits
  }
}

/**
 * Support for ES7+ where the `total` now is an object
 * @see https://www.elastic.co/guide/en/elasticsearch/reference/current/breaking-changes-7.0.html
 */
const getTotals = body => typeof body.hits.total === 'object' ? body.hits.total.value : body.hits.total

let esClient = null
function getClient (config) {
  if (esClient) {
    return esClient
  }
  
  let { host, port, protocol, apiVersion, requestTimeout, pingTimeout } = config.elasticsearch

  let nodes = []
  let hosts = typeof host === 'string' ? host.split(',') : host

  hosts.forEach(host => {
    const node = `${protocol}://${host}:${port}`
    nodes.push(node)
  })

  let auth
  if (config.elasticsearch.user) {
    const { user, password } = config.elasticsearch
    auth = { username: user, password }
  }

  esClient = new es.Client({ nodes, auth, apiVersion, requestTimeout, pingTimeout })

  return esClient
}

function putAlias (db, originalName, aliasName, next) {
  let step2 = () => {
    db.indices.putAlias({ index: originalName, name: aliasName }).then(result => {
      console.log('Index alias created')
    }).then(next).catch(err => {
      console.log(err.message)
      next()
    })
  }
  return db.indices.deleteAlias({
    index: aliasName,
    name: originalName
  }).then((result) => {
    console.log('Public index alias deleted')
    step2()
  }).catch((err) => {
    console.log('Public index alias does not exists', err.message)
    step2()
  })
}

function search (db, query) {
  return db.search(query)
}

function deleteIndex (db, indexName, next) {
  db.indices.delete({
    'index': indexName
  }).then((res) => {
    next()
  }).catch(err => {
    if (err) {
      console.error(err)
    }
    return db.indices.deleteAlias({
      index: '*',
      name: indexName
    }).then((result) => {
      console.log('Public index alias deleted')
      next()
    }).catch((err) => {
      console.log('Public index alias does not exists', err.message)
      next()
    })
  })
}

function reIndex (db, fromIndexName, toIndexName, next) {
  db.reindex({
    wait_for_completion: true,
    waitForCompletion: true,
    body: {
      'source': {
        'index': fromIndexName
      },
      'dest': {
        'index': toIndexName
      }
    }
  }).then(res => {
    next()
  }).catch(err => {
    next(err)
  })
}

/**
 * Load the schema definition for particular entity type
 * @param {String} entityType
 */
function loadSchema (entityType, apiVersion = '7.1') {
  const rootSchemaPath = path.join(__dirname, '../../config/elastic.schema.' + entityType + '.json')
  if (!fs.existsSync(rootSchemaPath)) {
    return null
  }
  let schemaContent = jsonFile.readFileSync(rootSchemaPath)
  let elasticSchema = parseInt(apiVersion) < 6 ? schemaContent : Object.assign({}, { mappings: schemaContent });
  const extensionsPath = path.join(__dirname, '../../config/elastic.schema.' + entityType + '.extension.json');
  if (fs.existsSync(extensionsPath)) {
    schemaContent = jsonFile.readFileSync(extensionsPath)
    let elasticSchemaExtensions = parseInt(apiVersion) < 6 ? schemaContent : Object.assign({}, { mappings: schemaContent });
    elasticSchema = _.merge(elasticSchema, elasticSchemaExtensions) // user extensions
  }
  return elasticSchema
}

function createIndex (db, indexName, collectionName, next) {
  let indexSchema = collectionName ? loadSchema(collectionName) : loadSchema('index', '5.6'); /** index schema is used only for 5.6 */

  const step2 = () => {
    db.indices.delete({
      'index': indexName
    }).then(res1 => {
      db.indices.create(
        {
          'index': indexName,
          'body': indexSchema
        }).then(res2 => {
        next()
      }).catch(err => {
        console.error(err)
        next(err)
      })
    }).catch(() => {
      db.indices.create(
        {
          'index': indexName,
          'body': indexSchema
        }).then(res2 => {
        next()
      }).catch(err => {
        console.error(err)
        next(err)
      })
    })
  }

  return db.indices.deleteAlias({
    index: '*',
    name: indexName
  }).then((result) => {
    console.log('Public index alias deleted')
    step2()
  }).catch((err) => {
    console.log('Public index alias does not exists', err.message)
    step2()
  })
}

// this is deprecated just for ES 5.6
function putMappings (db, indexName, next) {
  let productSchema = loadSchema('product', '5.6');
  let categorySchema = loadSchema('category', '5.6');
  let taxruleSchema = loadSchema('taxrule', '5.6');
  let attributeSchema = loadSchema('attribute', '5.6');
  let pageSchema = loadSchema('cms_page', '5.6');
  let blockSchema = loadSchema('cms_block', '5.6');

  Promise.all([
    db.indices.putMapping({
      index: indexName,
      type: 'product',
      body: productSchema
    }),
    db.indices.putMapping({
      index: indexName,
      type: 'taxrule',
      body: taxruleSchema
    }),
    db.indices.putMapping({
      index: indexName,
      type: 'attribute',
      body: attributeSchema
    }),
    db.indices.putMapping({
      index: indexName,
      type: 'cms_page',
      body: pageSchema
    }),
    db.indices.putMapping({
      index: indexName,
      type: 'cms_block',
      body: blockSchema
    }),
    db.indices.putMapping({
      index: indexName,
      type: 'category',
      body: categorySchema
    })
  ]).then(values => values.forEach(res => console.dir(res.body, { depth: null, colors: true })))
    .then(next)
    .catch(next)
}

module.exports = {
  putAlias,
  createIndex,
  deleteIndex,
  reIndex,
  search,
  adjustQuery,
  adjustQueryParams,
  adjustBackendProxyUrl,
  getClient,
  getHits,
  getTotals,
  adjustIndexName,
  putMappings
}
